storage/block/result.go

```diff
@@ -88,10 +98,78 @@ type fetchBlocksMetadataResult struct {
 	blocks []FetchBlockMetadataResult
 }
 
-// newFetchBlocksMetadataResult creates new database blocks metadata
-func newFetchBlocksMetadataResult(id string, blocks []FetchBlockMetadataResult) FetchBlocksMetadataResult {
+// NewFetchBlocksMetadataResult creates new database blocks metadata
+func NewFetchBlocksMetadataResult(id string, blocks []FetchBlockMetadataResult) FetchBlocksMetadataResult {
 	return fetchBlocksMetadataResult{id: id, blocks: blocks}
 }
 
 func (m fetchBlocksMetadataResult) ID() string                         { return m.id }
 func (m fetchBlocksMetadataResult) Blocks() []FetchBlockMetadataResult { return m.blocks }
+
+type filteredBlocksMetadataIter struct {
+	start     time.Time
+	end       time.Time
+	blockSize time.Duration
+	res       []FetchBlocksMetadataResult
+
+	id       string
+	metadata Metadata
+	ri       int
+	bi       int
+}
+
+// NewFilteredBlocksMetadataIter creates a new filtered blocks metadata iterator
+func NewFilteredBlocksMetadataIter(
+	start time.Time,
+	end time.Time,
+	blockSize time.Duration,
+	res []FetchBlocksMetadataResult,
+) FilteredBlocksMetadataIter {
+	return &filteredBlocksMetadataIter{
+		start:     start,
+		end:       end,
+		blockSize: blockSize,
+		res:       res,
+	}
+}
+
+func (it *filteredBlocksMetadataIter) Next() bool {
+	if it.ri >= len(it.res) {
+		return false
+	}
+	blocks := it.res[it.ri].Blocks()
+	for it.bi < len(blocks) {
+		block := blocks[it.bi]
+		if block.Err() != nil {
+			it.bi++
+			continue
```

Comment:
In that case, there will be no checksum for this peer, which means the number of hosts with checksums for this block is going to be smaller than the replica count, in which case we are going to log this block in the result.